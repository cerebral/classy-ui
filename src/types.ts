export interface IClass {
  id: string;
  classname: string;
  token: string;
  derived: string[] | null;
  variable: {
    value: string;
    originalValue: string;
  } | null;
}

export type IClasses = Record<string, IClass>;

export type ITokens = Record<string, IToken>;

export type IClassnames = Record<
  string,
  {
    tokens?:
      | ITokens
      | ((
          tokens: IGlobalTokens<IToken>,
          utils: {
            negative: (value: { [key: string]: IToken }) => { [key: string]: IToken };
          },
        ) => ITokens);
    css: ((name: string, value: string) => string) | string[];
    description?: string;
  }
>;

export interface IEvaluatedClassnames {
  [name: string]: {
    tokens: { [name: string]: IToken };
    tokensWithoutVariables: { [name: string]: IToken };
    css: ((name: string, value: string) => string) | string[];
    description?: string;
  };
}

export interface IToken {
  value: string;
  [key: string]: any;
}

export interface IGlobalTokens<T> {
  space: { [token: string]: T };
  fontSizes: { [token: string]: T };
  colors: { [token: string]: T };
  fonts: { [token: string]: T };
  fontWeights: { [token: string]: T };
  fontStyles: { [token: string]: T };
  lineHeights: { [token: string]: T };
  letterSpacings: { [token: string]: T };
  borderWidths: { [token: string]: T };
  radii: { [token: string]: T };
  size: { [token: string]: T };
  shadows: { [token: string]: T };
  zIndices: { [token: string]: T };

  durations: { [token: string]: T };
  timingFunctions: { [token: string]: T };
  opacity: { [token: string]: T };
  gridColumns: { [token: string]: T };
  gridTemplateColumns: { [token: string]: T };

  flexBases: { [token: string]: T };
  flexGrows: { [token: string]: T };
  flexShrinks: { [token: string]: T };
}

export interface IScreens {
  [key: string]: (css: string) => string;
}

export interface IConfig {
  tokens?: Partial<IGlobalTokens<string | IToken>>;
  screens?: {
    [name: string]: (css: string) => string;
  };
  themes?: {
    [name: string]: {
      [key in keyof IGlobalTokens<any>]?: Partial<IGlobalTokens<any>[key]>;
    };
  };
}

export interface IEvaluatedThemes {
  [tokens: string]: {
    [token: string]: {
      [theme: string]: string;
    };
  };
}

export interface IEvaluatedConfig {
  tokens: IGlobalTokens<IToken>;
  screens: {
    [name: string]: (css: string) => string;
  };
  classnames: IEvaluatedClassnames;
  themeNames: string[];
  themes: IEvaluatedThemes;
}

export interface IClassesByType {
  screens: {
    [type: string]: string[];
  };
  common: {
    [id: string]: string;
  };
  themeTokens: {
    [theme: string]: {
      [token: string]: string;
    };
  };
  rootTokens: {
    [token: string]: string;
  };
}

export interface IExtractedClass {
  id: string;
  name: string;
  decorators: string[];
  composition: string;
}

export interface IExtractedClasses {
  [composition: string]: {
    [id: string]: IExtractedClass;
  };
}
