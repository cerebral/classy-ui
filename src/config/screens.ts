import { IScreens } from '../types';

const screens: IScreens = {
  mobile: (css, breakpoints) => `@media (max-width: ${breakpoints.MOBILE}) {${css}}`,
  tablet: (css, breakpoints) => `@media (max-width: ${breakpoints.TABLET}) {${css}}`,
  laptop: (css, breakpoints) => `@media (max-width: ${breakpoints.LAPTOP}) {${css}}`,
  desktop: (css, breakpoints) => `@media (max-width: ${breakpoints.DESKTOP}) {${css}}`,
};

export default screens;
