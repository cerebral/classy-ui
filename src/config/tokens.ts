import { IGlobalTokens, IToken } from '../types';

const tokens: IGlobalTokens<string | IToken> = {
  fontStyles: {
    FONT_STYLE_NORMAL: 'normal',
    FONT_STYLE_ITALIC: 'italic',
    FONT_STYLE_OBLIQUE: 'oblique',
  },
  size: {
    NONE: '0',
    FULL: '100%',
    SCREEN_WIDTH: '100vw',
    SCREEN_HEIGHT: '100vh',
    SIZE_10: '20rem',
    SIZE_20: '24rem',
    SIZE_30: '28rem',
    SIZE_40: '32rem',
    SIZE_50: '36rem',
    SIZE_60: '42rem',
    SIZE_70: '48rem',
    SIZE_80: '56rem',
    SIZE_90: '64rem',
    SIZE_100: '72rem',
  },
  fontWeights: {
    WEIGHT_10: '100',
    WEIGHT_20: '200',
    WEIGHT_30: '300',
    WEIGHT_40: '400',
    WEIGHT_50: '500',
    WEIGHT_60: '600',
    WEIGHT_70: '700',
    WEIGHT_80: '800',
    WEIGHT_90: '900',
  },
  space: {
    SPACE_PX: '1px',
    SPACE_0: '0',
    SPACE_10: '0.25rem',
    SPACE_20: '0.5rem',
    SPACE_30: '0.75rem',
    SPACE_40: '1rem',
    SPACE_50: '1.25rem',
    SPACE_60: '1.5rem',
    SPACE_70: '2rem',
    SPACE_80: '2.5rem',
    SPACE_90: '3rem',
    SPACE_100: '4rem',
    SPACE_110: '5rem',
    SPACE_120: '6rem',
    SPACE_130: '8rem',
    SPACE_140: '10rem',
    SPACE_150: '12rem',
    SPACE_160: '14rem',
    SPACE_170: '16rem',
  },
  colors: {
    TRANSPARENT: 'transparent',
    BLACK: '#000',
    WHITE: '#fff',
    GRAY_10: '#f7fafc',
    GRAY_20: '#edf2f7',
    GRAY_30: '#e2e8f0',
    GRAY_40: '#cbd5e0',
    GRAY_50: '#a0aec0',
    GRAY_60: '#718096',
    GRAY_70: '#4a5568',
    GRAY_80: '#2d3748',
    GRAY_90: '#1a202c',
    RED_10: '#fff5f5',
    RED_20: '#fed7d7',
    RED_30: '#feb2b2',
    RED_40: '#fc8181',
    RED_50: '#f56565',
    RED_60: '#e53e3e',
    RED_70: '#c53030',
    RED_80: '#9b2c2c',
    RED_90: '#742a2a',
    ORANGE_10: '#fffaf0',
    ORANGE_20: '#feebc8',
    ORANGE_30: '#fbd38d',
    ORANGE_40: '#f6ad55',
    ORANGE_50: '#ed8936',
    ORANGE_60: '#dd6b20',
    ORANGE_70: '#c05621',
    ORANGE_80: '#9c4221',
    ORANGE_90: '#7b341e',
    YELLOW_10: '#fffff0',
    YELLOW_20: '#fefcbf',
    YELLOW_30: '#faf089',
    YELLOW_40: '#f6e05e',
    YELLOW_50: '#ecc94b',
    YELLOW_60: '#d69e2e',
    YELLOW_70: '#b7791f',
    YELLOW_80: '#975a16',
    YELLOW_90: '#744210',
    GREEN_10: '#f0fff4',
    GREEN_20: '#c6f6d5',
    GREEN_30: '#9ae6b4',
    GREEN_40: '#68d391',
    GREEN_50: '#48bb78',
    GREEN_60: '#38a169',
    GREEN_70: '#2f855a',
    GREEN_80: '#276749',
    GREEN_90: '#22543d',
    TEAL_10: '#e6fffa',
    TEAL_20: '#b2f5ea',
    TEAL_30: '#81e6d9',
    TEAL_40: '#4fd1c5',
    TEAL_50: '#38b2ac',
    TEAL_60: '#319795',
    TEAL_70: '#2c7a7b',
    TEAL_80: '#285e61',
    TEAL_90: '#234e52',
    BLUE_10: '#ebf8ff',
    BLUE_20: '#bee3f8',
    BLUE_30: '#90cdf4',
    BLUE_40: '#63b3ed',
    BLUE_50: '#4299e1',
    BLUE_60: '#3182ce',
    BLUE_70: '#2b6cb0',
    BLUE_80: '#2c5282',
    BLUE_90: '#2a4365',
    INDIGO_10: '#ebf4ff',
    INDIGO_20: '#c3dafe',
    INDIGO_30: '#a3bffa',
    INDIGO_40: '#7f9cf5',
    INDIGO_50: '#667eea',
    INDIGO_60: '#5a67d8',
    INDIGO_70: '#4c51bf',
    INDIGO_80: '#434190',
    INDIGO_90: '#3c366b',
    PURPLE_10: '#faf5ff',
    PURPLE_20: '#e9d8fd',
    PURPLE_30: '#d6bcfa',
    PURPLE_40: '#b794f4',
    PURPLE_50: '#9f7aea',
    PURPLE_60: '#805ad5',
    PURPLE_70: '#6b46c1',
    PURPLE_80: '#553c9a',
    PURPLE_90: '#44337a',
    PINK_10: '#fff5f7',
    PINK_20: '#fed7e2',
    PINK_30: '#fbb6ce',
    PINK_40: '#f687b3',
    PINK_50: '#ed64a6',
    PINK_60: '#d53f8c',
    PINK_70: '#b83280',
    PINK_80: '#97266d',
    PINK_90: '#702459',
  },
  fonts: {
    FONT_SANS: `${[
      'Inter',
      'apple-system',
      'BlinkMacSystemFont',
      '"Segoe UI"',
      'Roboto',
      '"Helvetica Neue"',
      'Arial',
      '"Noto Sans"',
      'sans-serif',
      '"Apple Color Emoji"',
      '"Segoe UI Emoji"',
      '"Segoe UI Symbol"',
      '"Noto Color Emoji"',
    ].join(', ')}`,
    FONT_SERIF: `${['Georgia', 'Cambria', '"Times New Roman"', 'Times', 'serif'].join(',')}`,
    FONT_MONO: `${['Menlo', 'Monaco', 'Consolas', '"Liberation Mono"', '"Courier New"', 'monospace'].join(',')}`,
  },
  borderWidths: {
    WIDTH_0: '0',
    WIDTH_1: '1px',
    WIDTH_2: '2px',
    WIDTH_4: '4px',
    WIDTH_8: '8px',
  },
  letterSpacings: {
    SPACING_10: '-0.05em',
    SPACING_20: '-0.025em',
    SPACING_30: '0',
    SPACING_40: '0.025em',
    SPACING_50: '0.05em',
    SPACING_60: '0.1em',
  },
  radii: {
    RADIUS_0: '0',
    RADIUS_10: '.125rem',
    RADIUS_20: '.25rem',
    RADIUS_30: '.5rem',
    RADIUS_FULL: '9999px',
  },
  lineHeights: {
    HEIGHT_10: '1',
    HEIGHT_20: '1.25',
    HEIGHT_30: '1.375',
    HEIGHT_40: '1.5',
    HEIGHT_50: '1.625',
    HEIGHT_60: '2',
  },
  shadows: {
    SHADOW_10: '0 0 0 1px rgba(0, 0, 0, 0.05)',
    SHADOW_20: '0 1px 2px 0 rgba(0, 0, 0, 0.05)',
    SHADOW_30: '0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)',
    SHADOW_40: '0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)',
    SHADOW_50: '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)',
    SHADOW_60: '0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)',
    SHADOW_70: '0 25px 50px -12px rgba(0, 0, 0, 0.25)',
    SHADOW_INNER: 'inset 0 2px 4px 0 rgba(0, 0, 0, 0.06)',
    SHADOW_OUTLINE: '0 0 0 3px rgba(66, 153, 225, 0.5)',
    SHADOW_NONE: 'none',
  },
  zIndices: {
    Z_0: '0',
    Z_1: '1',
    Z_2: '2',
    Z_3: '3',
    Z_4: '4',
    Z_5: '5',
    Z_6: '6',
    Z_7: '7',
    Z_8: '8',
    Z_9: '9',
  },
  opacity: {
    OPACITY_0: '0',
    OPACITY_25: '0.25',
    OPACITY_50: '0.5',
    OPACITY_75: '0.75',
    OPACITY_FULL: '1',
  },
  durations: {
    DURATION_10: '75ms',
    DURATION_20: '100ms',
    DURATION_30: '150ms',
    DURATION_40: '200ms',
    DURATION_50: '300ms',
    DURATION_60: '500ms',
    DURATION_70: '700ms',
    DURATION_80: '1000ms',
  },
  timingFunctions: {
    LINEAR: 'linear',
    IN: 'cubic-bezier(0.4, 0, 1, 1)',
    OUT: 'cubic-bezier(0, 0, 0.2, 1)',
    IN_OUT: 'cubic-bezier(0.4, 0, 0.2, 1)',
  },
  fontSizes: {
    SIZE_10: '0.75rem',
    SIZE_20: '0.875rem',
    SIZE_30: '1rem',
    SIZE_40: '1.125rem',
    SIZE_50: '1.25rem',
    SIZE_60: '1.5rem',
    SIZE_70: '1.875rem',
    SIZE_80: '2.25rem',
    SIZE_90: '3rem',
    SIZE_100: '4rem',
  },
  gridColumns: {
    AUTO: 'auto',
    SPAN_1: 'span 1',
    SPAN_2: 'span 2',
    SPAN_3: 'span 3',
    SPAN_4: 'span 4',
    SPAN_5: 'span 5',
    SPAN_6: 'span 6',
    SPAN_7: 'span 7',
    SPAN_8: 'span 8',
    SPAN_9: 'span 9',
    SPAN_10: 'span 10',
    SPAN_11: 'span 11',
    SPAN_12: 'span 12',
  },
  gridTemplateColumns: {
    COLUMNS_NONE: 'none',
    COLUMNS_1: 'repeat(1, minmax(0, 1fr))',
    COLUMNS_2: 'repeat(2, minmax(0, 1fr))',
    COLUMNS_3: 'repeat(3, minmax(0, 1fr))',
    COLUMNS_4: 'repeat(4, minmax(0, 1fr))',
    COLUMNS_5: 'repeat(5, minmax(0, 1fr))',
    COLUMNS_6: 'repeat(6, minmax(0, 1fr))',
    COLUMNS_7: 'repeat(7, minmax(0, 1fr))',
    COLUMNS_8: 'repeat(8, minmax(0, 1fr))',
    COLUMNS_9: 'repeat(9, minmax(0, 1fr))',
    COLUMNS_10: 'repeat(10, minmax(0, 1fr))',
    COLUMNS_11: 'repeat(11, minmax(0, 1fr))',
    COLUMNS_12: 'repeat(12, minmax(0, 1fr))',
  },
  flexBases: {
    AUTO: 'auto',
    FILL: 'fill',
    MAX_CONTENT: 'max-content',
    MIN_CONTENT: 'min-content',
    FIT_CONTENT: 'fit-content',
    CONTENT: 'content',
    BASIS_0: '0',
    BASIS_10: '0.75rem',
    BASIS_20: '0.875rem',
    BASIS_30: '1rem',
    BASIS_40: '1.125rem',
    BASIS_50: '1.25rem',
    BASIS_60: '1.5rem',
    BASIS_70: '1.875rem',
    BASIS_80: '2.25rem',
    BASIS_90: '3rem',
    BASIS_100: '4rem',
    BASIS_110: '5rem',
    BASIS_120: '6rem',
    BASIS_130: '8rem',
    BASIS_140: '10rem',
    BASIS_150: '12rem',
    BASIS_160: '14rem',
    BASIS_170: '16rem',
  },
  flexGrows: {
    GROW_0: '0',
    GROW_10: '0.1',
    GROW_20: '0.2',
    GROW_30: '0.3',
    GROW_40: '0.4',
    GROW_50: '0.5',
    GROW_60: '0.6',
    GROW_70: '0.7',
    GROW_80: '0.8',
    GROW_90: '0.9',
    GROW_100: '1',
    GROW_110: '1.1',
    GROW_120: '1.2',
    GROW_130: '1.3',
    GROW_140: '1.4',
    GROW_150: '1.5',
    GROW_160: '1.6',
    GROW_170: '1.7',
    GROW_180: '1.8',
    GROW_190: '1.9',
    GROW_200: '2',
    GROW_210: '2.1',
    GROW_220: '2.2',
    GROW_230: '2.3',
    GROW_240: '2.4',
    GROW_250: '2.5',
    GROW_260: '2.6',
    GROW_270: '2.7',
    GROW_280: '2.8',
    GROW_290: '2.9',
    GROW_300: '3',
  },
  flexShrinks: {
    SHRINK_0: '0',
    SHRINK_10: '0.1',
    SHRINK_20: '0.2',
    SHRINK_30: '0.3',
    SHRINK_40: '0.4',
    SHRINK_50: '0.5',
    SHRINK_60: '0.6',
    SHRINK_70: '0.7',
    SHRINK_80: '0.8',
    SHRINK_90: '0.9',
    SHRINK_100: '1',
    SHRINK_110: '1.1',
    SHRINK_120: '1.2',
    SHRINK_130: '1.3',
    SHRINK_140: '1.4',
    SHRINK_150: '1.5',
    SHRINK_160: '1.6',
    SHRINK_170: '1.7',
    SHRINK_180: '1.8',
    SHRINK_190: '1.9',
    SHRINK_200: '2',
  },
};

export default tokens;
