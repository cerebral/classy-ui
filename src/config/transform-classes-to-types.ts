import { IEvaluatedConfig } from '../types';
import { allowedPseudoDecorators, camelToDash, getScreens } from '../utils';

function convertClassnameToType(baseClass: string, token: string, config: IEvaluatedConfig) {
  const cssConstructor = config.classnames[baseClass].css;
  const css: string[] =
    typeof cssConstructor === 'function'
      ? [
          cssConstructor(`.${camelToDash(baseClass)}-${token}`, config.classnames[baseClass].tokens[token].value)
            .split('\n')
            .join('\n   * '),
        ]
      : cssConstructor.reduce<string[]>((aggr, derivedBaseClass) => {
          return aggr.concat(
            (config.classnames[derivedBaseClass].css as any)(
              `.${camelToDash(baseClass)}-${token}`,
              config.classnames[derivedBaseClass].tokens[token].value,
            )
              .split('\n')
              .join('\n   * '),
          );
        }, []);
  return `
  /**
   ${
     config.classnames[baseClass].tokens[token].value.startsWith('#')
       ? `* ![${config.classnames[baseClass].tokens[token]}](https://placehold.it/15/${config.classnames[
           baseClass
         ].tokens[token].value.substr(1)}/000000?text=+)`
       : ''
   }
   ${Object.keys(config.classnames[baseClass].tokens[token]).reduce((aggr, key) => {
     if (key === 'value') {
       return aggr;
     }

     return `${aggr}* **${key}**: ${config.classnames[baseClass].tokens[token][key]}\n`;
   }, '')}
   * \`\`\`css
   * ${css.join('\n')}
   * \`\`\`
   */
  readonly ${token}: IDecorators;
  `;
}

export const transform = (config: IEvaluatedConfig) => {
  return `
declare module 'classy-ui' {
  export type TClassname = string & "CLASSNAME";
  export enum Themes {
    ${config.themeNames
      .reduce<string[]>((aggr, theme) => aggr.concat(`${theme} = "${`themes-${theme}"`}`), [])
      .join(',\n')}
  }
  export type TGroup = 'GROUP'
  export interface IDecorators {
    /**
     * \`\`\`css
     * group:hover $token {}
     * \`\`\`
     */
    groupHover: IDecorators;
    /**
     * \`\`\`css
     * group:focus $token {}
     * \`\`\`
     */
    groupFocus: IDecorators;
    /**
     * \`\`\`css
     * group:active $token {}
     * \`\`\`
     */
    groupActive: IDecorators;
    /**
     * \`\`\`css
     * group:first-child $token {}
     * \`\`\`
     */
    groupFirstChild: IDecorators;
    /**
     * \`\`\`css
     * group:last-child $token {}
     * \`\`\`
     */
    groupLastChild: IDecorators;
    /**
     * \`\`\`css
     * group:odd-child $token {}
     * \`\`\`
     */
    groupOddChild: IDecorators;
    /**
     * \`\`\`css
     * group:even-child $token {}
     * \`\`\`
     */
    groupEvenChild: IDecorators;
    /**
     * \`\`\`css
     * group:focus-within $token {}
     * \`\`\`
     */
    groupFocusWithin: IDecorators;
  ${allowedPseudoDecorators
    .map(
      decorator => `
    /**
     * \`\`\`css
     * $token:${camelToDash(decorator)} {}
     * \`\`\`
     */
    ${decorator}: IDecorators;
  `,
    )
    .join('\n')}
  }
  export type TCompose = (...args: Array<IDecorators | TGroup | Themes | TClassname | boolean | ''>) => TClassname;
  export type TTokens = { 
    ${Object.keys(config.classnames)
      .reduce<string[]>((aggr, baseClass) => {
        return aggr.concat(`
    ${
      config.classnames[baseClass].description
        ? `
  /**
    * ${config.classnames[baseClass].description}
    */`
        : ''
    }
    ${baseClass}: {
      ${Object.keys(config.classnames[baseClass].tokens)
        .reduce<string[]>((subAggr, token) => {
          return subAggr.concat(convertClassnameToType(baseClass, token, config));
        }, [])
        .join('\n')}
    }
  `);
      }, [])
      .join('\n')}
  }

  export const themes: {
    ${config.themeNames.reduce<string[]>((aggr, theme) => aggr.concat(`${theme}: Themes.${theme}`), []).join('\n')}
  };
  export const group: TGroup;
  export const compose: TCompose;
  export const tokens: TTokens;
  export const c: TCompose;
  export const t: TTokens;

  ${Object.keys(config.screens)
    .map(
      screen => `
    /**
     * \`\`\`css
     * \\${config.screens[screen](`\n      $token\n`, getScreens(config))}
     * \`\`\`
     */
    export const ${screen}: TCompose;
  `,
    )
    .join('\n')}
}`;
};
