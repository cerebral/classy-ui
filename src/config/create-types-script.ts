import { writeFileSync } from 'fs';
import { join } from 'path';

import { evaluateConfig } from '../utils';

import { transform as transformClassesToTypes } from './transform-classes-to-types';

const config = evaluateConfig({});

if (process.env.NODE_ENV !== 'test') {
  const esTypesPath = join(process.cwd(), 'es', 'classy-ui.d.ts');
  const libTypesPath = join(process.cwd(), 'lib', 'classy-ui.d.ts');
  const types = transformClassesToTypes(config);
  writeFileSync(esTypesPath, types);
  writeFileSync(libTypesPath, types);
}
