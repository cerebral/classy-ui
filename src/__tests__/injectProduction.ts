import { transform } from '../config/transform-config-to-classes';
import { IClassesByType, IExtractedClasses } from '../types';
import {
  createExtractedClasses,
  createProductionClassObjects,
  createProductionCss,
  evaluateConfig,
  getUserConfig,
  injectProduction,
} from '../utils';

// tslint:disable-next-line
const config = evaluateConfig(getUserConfig());
const classes = transform(config);

describe('INJECT PRODUCTION', () => {
  test('should inject simple', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with pseudo selector', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'color', token: 'RED', decorators: ['hover'] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with breakpoint', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject both pseudo selector and breakpoint', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'RED', decorators: ['hover'] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with multiple pseudo selectors', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'color', token: 'RED', decorators: ['hover', 'firstChild'] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject multiple', () => {
    const evaluatedProductionShortnames = {
      classnames: [],
      tokens: [],
      decorators: [],
    };
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        evaluatedProductionShortnames,
      ).concat(
        createProductionClassObjects(
          { composition: 'tablet', baseClass: 'color', token: 'BLUE', decorators: ['hover'] },
          classes,
          evaluatedProductionShortnames,
        ),
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject themes', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'GREEN', decorators: [] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group breakpoints', () => {
    const evaluatedProductionShortnames = {
      classnames: [],
      tokens: [],
      decorators: [],
    };
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        evaluatedProductionShortnames,
      ).concat(
        createProductionClassObjects(
          { composition: 'tablet', baseClass: 'color', token: 'BLUE', decorators: [] },
          classes,
          evaluatedProductionShortnames,
        ),
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group themes', () => {
    const evaluatedProductionShortnames = {
      classnames: [],
      tokens: [],
      decorators: [],
    };
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'color', token: 'GREEN', decorators: [] },
        classes,
        evaluatedProductionShortnames,
      ).concat(
        createProductionClassObjects(
          { composition: 'compose', baseClass: 'color', token: 'PURPLE', decorators: [] },
          classes,
          evaluatedProductionShortnames,
        ),
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group themes and breakpoints', () => {
    const evaluatedProductionShortnames = {
      classnames: [],
      tokens: [],
      decorators: [],
    };
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'tablet', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        evaluatedProductionShortnames,
      ).concat(
        createProductionClassObjects(
          { composition: 'tablet', baseClass: 'color', token: 'PURPLE', decorators: [] },
          classes,
          evaluatedProductionShortnames,
        ),
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should handle references classnames', () => {
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'borderColor', token: 'GREEN', decorators: ['hover'] },
        classes,
        {
          classnames: [],
          tokens: [],
          decorators: [],
        },
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should not override with screens', () => {
    const evaluatedProductionShortnames = {
      classnames: [],
      tokens: [],
      decorators: [],
    };
    const classCollection: IExtractedClasses = createExtractedClasses(
      createProductionClassObjects(
        { composition: 'compose', baseClass: 'color', token: 'RED', decorators: [] },
        classes,
        evaluatedProductionShortnames,
      ).concat(
        createProductionClassObjects(
          { composition: 'compose', baseClass: 'color', token: 'BLUE', decorators: ['mobile'] },
          classes,
          evaluatedProductionShortnames,
        ),
      ),
    );

    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
});
