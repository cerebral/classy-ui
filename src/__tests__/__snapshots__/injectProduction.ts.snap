// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INJECT PRODUCTION should group breakpoints 1`] = `
"@media (max-width: 768px) {.A-A_A {
  color: red;
}.A-A_B {
  color: blue;
}}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should group themes 1`] = `
".A_A {
  color: var(--colors-GREEN);
}.A_B {
  color: var(--colors-PURPLE);
}:root{--colors-GREEN:green;--colors-PURPLE:purple;}.themes-dark{--colors-GREEN:yellow;--colors-PURPLE:pink;}"
`;

exports[`INJECT PRODUCTION should group themes and breakpoints 1`] = `
"@media (max-width: 768px) {.A-A_A {
  color: red;
}.A-A_B {
  color: var(--colors-PURPLE);
}}:root{--colors-PURPLE:purple;}.themes-dark{--colors-PURPLE:pink;}"
`;

exports[`INJECT PRODUCTION should handle references classnames 1`] = `
".A-A_A:hover {
  border-top-color: var(--colors-GREEN);
}.A-B_A:hover {
  border-right-color: var(--colors-GREEN);
}.A-C_A:hover {
  border-bottom-color: var(--colors-GREEN);
}.A-D_A:hover {
  border-left-color: var(--colors-GREEN);
}:root{--colors-GREEN:green;}.themes-dark{--colors-GREEN:yellow;}"
`;

exports[`INJECT PRODUCTION should inject both pseudo selector and breakpoint 1`] = `
"@media (max-width: 768px) {.AB-A_A:hover {
  color: red;
}}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should inject multiple 1`] = `
"@media (max-width: 768px) {.A-A_A {
  color: red;
}.BA-A_B:hover {
  color: blue;
}}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should inject simple 1`] = `
".A_A {
  color: red;
}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should inject themes 1`] = `
"@media (max-width: 768px) {.A-A_A {
  color: var(--colors-GREEN);
}}:root{--colors-GREEN:green;}.themes-dark{--colors-GREEN:yellow;}"
`;

exports[`INJECT PRODUCTION should inject with breakpoint 1`] = `
"@media (max-width: 768px) {.A-A_A {
  color: red;
}}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should inject with multiple pseudo selectors 1`] = `
".AB-A_A:hover:first-child {
  color: red;
}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should inject with pseudo selector 1`] = `
".A-A_A:hover {
  color: red;
}.themes-dark{}"
`;

exports[`INJECT PRODUCTION should not override with screens 1`] = `
".A_A {
  color: red;
}.A-A_B {
  color: blue;
}.themes-dark{}"
`;
